opnode/l2/api.go

```diff
@@ -0,0 +1,206 @@
+	return hexutil.Encode(b[:])
+}
+
+type Uint64Quantity = hexutil.Uint64
+
+type BytesMax32 []byte
+
+func (b *BytesMax32) UnmarshalJSON(text []byte) error {
+	if len(text) > 64+2+2 { // account for delimiter "", and 0x prefix
+		return fmt.Errorf("input too long, expected at most 32 hex-encoded, 0x-prefixed, bytes: %x", text)
+	}
+	return (*hexutil.Bytes)(b).UnmarshalJSON(text)
+}
+
+func (b *BytesMax32) UnmarshalText(text []byte) error {
+	if len(text) > 64+2 { // account for 0x prefix
+		return fmt.Errorf("input too long, expected at most 32 hex-encoded, 0x-prefixed, bytes: %x", text)
+	}
+	return (*hexutil.Bytes)(b).UnmarshalText(text)
+}
+
+func (b BytesMax32) MarshalText() ([]byte, error) {
+	return (hexutil.Bytes)(b).MarshalText()
+}
+
+func (b BytesMax32) String() string {
+	return hexutil.Encode(b)
+}
+
+type Uint256Quantity = uint256.Int
+
+type Data = hexutil.Bytes
+
+// TODO: implement neat 8 byte typed payload ID and upstream it to geth api definitions
+type PayloadID = hexutil.Bytes
+
+type ExecutionPayload struct {
+	ParentHashField common.Hash     `json:"parentHash"`
+	FeeRecipient    common.Address  `json:"feeRecipient"`
+	StateRoot       Bytes32         `json:"stateRoot"`
+	ReceiptsRoot    Bytes32         `json:"receiptsRoot"`
+	LogsBloom       Bytes256        `json:"logsBloom"`
+	Random          Bytes32         `json:"random"`
+	BlockNumber     Uint64Quantity  `json:"blockNumber"`
+	GasLimit        Uint64Quantity  `json:"gasLimit"`
+	GasUsed         Uint64Quantity  `json:"gasUsed"`
+	Timestamp       Uint64Quantity  `json:"timestamp"`
+	ExtraData       BytesMax32      `json:"extraData"`
+	BaseFeePerGas   Uint256Quantity `json:"baseFeePerGas"`
+	BlockHash       common.Hash     `json:"blockHash"`
+	// Array of transaction objects, each object is a byte list (DATA) representing
+	// TransactionType || TransactionPayload or LegacyTransaction as defined in EIP-2718
+	TransactionsField []Data `json:"transactions"`
+}
+
+func (payload *ExecutionPayload) ID() eth.BlockID {
+	return eth.BlockID{Hash: payload.BlockHash, Number: uint64(payload.BlockNumber)}
+}
+
+// Implement block interface to enable derive.BlockReferences over a payload
+// type Block interface {
+// 	Hash() common.Hash
+// 	NumberU64() uint64
+// 	ParentHash() common.Hash
+// 	Transactions() types.Transactions
+// }
+
+func (payload *ExecutionPayload) Hash() common.Hash {
+	return payload.BlockHash
+}
+
+func (payload *ExecutionPayload) NumberU64() uint64 {
+	return uint64(payload.BlockNumber)
+}
+
+func (payload *ExecutionPayload) Time() uint64 {
+	return uint64(payload.Timestamp)
+}
+
+func (payload *ExecutionPayload) ParentHash() common.Hash {
+	return payload.ParentHashField
+}
+
+func (payload *ExecutionPayload) Transactions() types.Transactions {
+	res := make([]*types.Transaction, len(payload.TransactionsField))
+	for i, t := range payload.TransactionsField {
+		res[i] = new(types.Transaction)
+		err := res[i].UnmarshalBinary(t)
+		if err != nil {
+			panic(err)
```

Comment:
I think this `panic` may be problematic. The `ExecutionPayload` struct represents its transactions list as a list of `HexBytes`, so a struct containing invalid byte sequences could be unmarshaled without error only to eventually crash the program when this method is called.

If we always completely trust the L2 geth instance this might not be a problem, but IMO we should avoid panicking purely for the sake of re-using an interface method.

Reply:
Agreed, this `panic` is dirty

Reply:
I think it's like this because it tries to emulate the same `derive.Block` interface as the geth `types.Block` that returns parsed transactions.

Reply:
Part of the problem here is that our current `l2.Engine` is a mix of the Engine and Eth RPC namespaces, without any caching or minimal block/header types like we do in `l1.Source`. I'd say we keep this as-is for now, but solve it as part of refactoring/improving the L2 source.